lib/IR/GraphScheduler.cpp

```diff
@@ -170,6 +188,48 @@ void ChildMemSizeBasedScheduler::schedule() {
   scheduleNodes();
 }
 
+void TopologicalSortBasedScheduler::schedule() {
+  for (auto &N : G_.getNodes()) {
+    indegree_[&N] = 0;
+  }
+
+  for (auto &N : G_.getNodes()) {
+    for (auto &use : N.getUsers()) {
+      indegree_[use.getUser()]++;
+    }
+  }
+
+  std::queue<Node *> q;
+  for (auto &N : G_.getNodes()) {
+    if (indegree_[&N] == 0) {
+      q.push(&N);
+    }
+  }
+
+  while (!q.empty()) {
+    Node *cur = q.front();
+    scheduled_.push_back(cur);
+    q.pop();
+    for (auto &use : cur->getUsers()) {
+      indegree_[use.getUser()]--;
+      if (indegree_[use.getUser()] == 0) {
+        q.push(use.getUser());
+      }
+    }
+  }
+}
+
+Scheduler *createScheduler(SchedulerKind schedulerKind, Function &G,
+                           NodesPtrList &scheduled) {
+  switch (schedulerKind) {
+  case SchedulerKind::ChildMemSizeBased:
+    return new ChildMemSizeBasedScheduler(G, scheduled);
```

Comment:
can you move schedulers into its own files?

Reply:
Not very clear about this, do you mean spilit GraphScheduler.cpp into two files, ChildMemSizeBasedScheduler.cpp and TopologicalSortBasedScheduler.cpp?

Reply:
So the file structure will be 

```
IR---
    ChildMemSizeBasedScheduler.cpp
    CMakeLists.txt
    GraphScheduler.cpp
    GraphScheduler.h
    Instrs.cpp
    IR.cpp
    IRBuilder.cpp
    IRGen.cpp
    IRUtils.cpp
    TopologicalSortBasedScheduler.cpp
```

Or do we need another directory `GraphScheduler`?

```
IR---
    CMakeLists.txt
    GraphScheduler.cpp
    GraphScheduler.h
    GraphScheduler---
         ChildMemSizeBasedScheduler.h
         ChildMemSizeBasedScheduler.cpp
         TopologicalSortBasedScheduler.h
         TopologicalSortBasedScheduler.cpp
    Instrs.cpp
    IR.cpp
    IRBuilder.cpp
    IRGen.cpp
    IRUtils.cpp
```

@rdzhabarov @nadavrot 
Which one do you prefer?

Reply:
The same folder is fine, we always can add that if needed later.